src/rulesets/Base/mapreduce.jl

```diff
@@ -322,3 +325,122 @@ end
     end
     return dx
 end
+
+#####
+##### `foldl`
+#####
+
+# `foldl` guarantees to execute `f` in order, left to right. So it makes sense even when
+# this `f` is stateful, in which case the gradient must be calculated in the reverse order. 
+# The implementation aims to be efficient for both tuples and arrays. 
+# Note that it does not return a gradient for `init`.
+
+function rrule(
+        config::RuleConfig{>:HasReverseMode}, ::typeof(foldl), f::F, x::Union{AbstractArray, Tuple};
+        init=_InitialValue()
+    ) where {F}
+    list, start = if init !== _InitialValue()
+        x, init
+    else
+        _drop1(x), first(x)
+    end
+    hobbits = accumulate(list; init=(start, pi)) do (a,_), b
+        c, back = rrule_via_ad(config, f, a, b)  # LHS is just documentation here
+        # Don't need to store every `c`, need last + one for the next iteration
+    end
+    y = first(last(hobbits))
+    axe = x isa AbstractArray ? axes(x) : nothing
+    type = Val(typeof(x))
+    project = x isa AbstractArray{<:Number} ? ProjectTo(x) : identity
+    function unfoldl(dy)
+        _tini = (pi, unthunk(dy), pi)
```

Comment:
Why are their `pi`s here?

Reply:
Because at some point, some lines got too long with `nothing`. The point is that these aren't used. This could be made some special struct with no methods, to explicitly check that it's not used. Or just `0` if we trust the tests.

Reply:
But why are they put in then ignored?

Reply:
Because you have to put something. The same function is applied to every iteration, and so `init` must match the signature it's expecting on subsequent iterations. Each iteration here writes 3 values, only one of which is read by the next step.